### DISCUSSION TITLE
Is it possible to use env variable (secrets) in a plugin?

### DISCUSSION THREAD

**Discussion by EverettDunamu** [Kong/kong#10538] [Help]
I want to store my api key secrets in a safe place and load them to inject the secret values in a plugin called `Request Transformer`. Actually, I want my kong gateway can access external APIs without me passing api key credentials every time.

So I found that the request transformer can add request header when I request to a kong gateway. But the request header values are stored in the DB in plaintext format. I don't want the credentials are stored in my DB in plaintext format. So I want to know who to store credentials and load them into plugin values. Is there any way to do this?

For example, there is an external server `www.external.com/api/v1/hello` and it accepts requests with `x-api-key: top-secret` header. Then assume that I set a route `www.my-kong.com/my-route` to a service `www.external.com/v1/hello`. After that, I want to request to my kong gateway `www.my-kong.com/my-route` without `x-api-key` header. I want my kong gateway manage and handle the credentials instead of me.

**Comment by ghost** [comment]
I think the Keyring and Data Encryption feature in Kong Enterprise Edition seems to be what you need.
read more: https://docs.konghq.com/gateway/latest/kong-enterprise/db-encryption/

**Comment by robhayesmba** [comment]
If you are looking at the Kong Enterprise Edition, you may want to look at the Secrets Management feature: https://docs.konghq.com/gateway/latest/kong-enterprise/secrets-management/. According to the documentation it allows environment variables as a backend among others.

Have not used it myself.  Have it on my to-do list to look and test out.
